### PR TITLE
Dynamic source loading

### PR DESCRIPTION
**Pull request by matzemathics**
Only open the source files when they are actually needed.

### CODE REVIEW

**Comment by monsterkrampe** [comment]
I appreciate the refactoring!
From skimming through, I fail to see though how this PR changes the behavior into:
>Only open the source files when they are actually needed.


Wasn't this the case before already? (Maybe I'm really just not seeing it.)

**Comment by matzemathics** [comment]
> I appreciate the refactoring! From skimming through, I fail to see though how this PR changes the behavior into:
> 
> > Only open the source files when they are actually needed.
> 
> Wasn't this the case before already? (Maybe I'm really just not seeing it.)

Yes, you are right (I just double checked). So the only big change is that the `TableReader` trait now consumes the reader when reading. I guess that also makes the `NativeDataSourceResolver` sort of obsolete.

File: nemo-physical/src/table_reader.rs [hunk 1]
```diff
@@ -4,7 +4,7 @@
 use crate::{builder_proxy::PhysicalBuilderProxyEnum, error::ReadingError};
 
 /// Resource that can be reference in source declarations in Nemo programs
-/// Resources are resolved using `nemo::io::input_manager::ResourceProviders`
+/// Resources are resolved using `nemo::io::resource_providers::ResourceProviders`
```
Comment (mmarx) [inline nemo-physical/src/table_reader.rs:7]: I guess this should be a link (if it were, we would have gotten a warning already for the target not existing)

File: nemo/src/model/rule_model/datasource.rs [hunk 2]
```diff
@@ -215,7 +215,7 @@ impl DataSource for SparqlQuery {
 
 /// An external data source.
 #[derive(Debug, Clone, PartialEq, Eq)]
-pub enum DataSourceT {
+pub enum NativeDataSource {
```
Comment (mmarx) [inline nemo/src/model/rule_model/datasource.rs:218]: What would the non-native data sources be that we want to support?

Comment (matzemathics) [reply nemo/src/model/rule_model/datasource.rs:218]: I'd like to build some integration to input python tuples into the database, without going through the resource provider trait (which only works for serialized data).

**Review by monsterkrampe — APPROVED** [review]

**Pull request merged**